hal/common/retarget.cpp

```diff
@@ -477,24 +478,80 @@ extern "C" WEAK void __cxa_pure_virtual(void) {
 
 #endif
 
+static uint32_t current_size = 0;
+static uint32_t max_size = 0;
+
+uint32_t mbed_stats_heap_max()
+{
+    return max_size;
+}
+
+uint32_t mbed_stats_heap_cur()
+{
+    return current_size;
+}
+
 #if defined(TOOLCHAIN_GCC)
 #ifdef   FEATURE_UVISOR
 #include "uvisor-lib/uvisor-lib.h"
 #endif/* FEATURE_UVISOR */
 
 #ifndef  FEATURE_UVISOR
 extern "C" {
+
+extern "C" void __malloc_lock( struct _reent *_r );
+extern "C" void __malloc_unlock( struct _reent *_r );
+
 void * __wrap__malloc_r(struct _reent * r, size_t size) {
     extern void * __real__malloc_r(struct _reent * r, size_t size);
-    return __real__malloc_r(r, size);
+
+    __malloc_lock(r);
+    uint32_t *data = (uint32_t*)__real__malloc_r(r, size + 4);
```

Comment:
perhaps this functionaly should be behind some compiler flag because now this reserve 4 extra bytes every single malloc,, at least this should be documented because there is devices with very little amount of memory..